### PR TITLE
metal: implement get_timestamp_period

### PR DESCRIPTION
**Pull request by jinleili**
**Description**
Implemented by referring to [Apple's Correlating CPU and GPU Timestamps](https://developer.apple.com/documentation/metal/counter_sampling/correlating_cpu_and_gpu_timestamps?language=objc)

[Metal `sampleTimestamps:gpuTimestamp:` API ](https://developer.apple.com/documentation/metal/mtldevice/3194378-sampletimestamps?language=occ)available on iOS | iPadOS 14 and macOS 10.15+, but on old A8 chip, the timestamp value obtained by this API is 0.

**Testing**
On all tested Apple silicon (A-series + M1), obtained cpu_timestamp equals to gpu_timestamp:

On iPad mini4(iPadOS 14.7.1),  `sampleTimestamps: gpuTimestamp:` alway get 0

| Set | cpu_timestamp0 | cpu_timestamp1 | gpu_timestamp0 | gpu_timestamp1 | cpu_delta / gpu_delta |
|---|---|---|---|---|---|
|0|0|0|0|0|NaN|
|1|0|0|0|0|NaN|
|2|0|0|0|0|NaN|
|3|0|0|0|0|NaN|
|4|0|0|0|0|NaN|

M1 Mac `get_timestamp_period` alway equal 1.0

| Set | cpu_timestamp0 | cpu_timestamp1 | gpu_timestamp0 | gpu_timestamp1 | cpu_delta / gpu_delta |
|---|---|---|---|---|---|
| 0 | 138194481157083 | 138194481157458 | 138194481157083 | 138194481157458 | 1.0 |
|1|138255790533166|138255790533500|138255790533166|138255790533500|1.0|
|2|138287026088250|138287026088458|138287026088250|138287026088458|1.0|
|3|138316982792458|138316982792791|138316982792458|138316982792791|1.0|
|4|138347604571125|138347604571416|138347604571125|138347604571416|1.0|

iPad Pro 2018  `get_timestamp_period` alway equal 1.0

| Set | cpu_timestamp0 | cpu_timestamp1 | gpu_timestamp0 | gpu_timestamp1 | cpu_delta / gpu_delta |
|---|---|---|---|---|---|
|0|6877203414500|6877203415209|6877203414500|6877203415209|1.0|
|1|6882077772500|6882077773167|6882077772500|6882077773167|1.0|
|2|6888188876209|6888188877000|6888188876209|6888188877000|1.0|
|3|6893635111125|6893635111834|6893635111125|6893635111834|1.0|
|4|6898661053167|6898661053834|6898661053167|6898661053834|1.0|

iPhone 12 Pro  `get_timestamp_period` alway equal 1.0

| Set | cpu_timestamp0 | cpu_timestamp1 | gpu_timestamp0 | gpu_timestamp1 | cpu_delta / gpu_delta |
|---|---|---|---|---|---|
|0|1213782040268791|1213782040269208|1213782040268791|1213782040269208|1.0|
|1|1213790290374500|1213790290374833|1213790290374500|1213790290374833|1.0|
|2|1213796168203083|1213796168203416|1213796168203083|1213796168203416|1.0|
|3|1213801947666250|1213801947666375|1213801947666250|1213801947666375|1.0|
|4|1213806890532208|1213806890532666|1213806890532208|1213806890532666|1.0|

Intel Mac with UHD Graphics 630

| Set | cpu_timestamp0 | cpu_timestamp1 | gpu_timestamp0 | gpu_timestamp1 | cpu_delta / gpu_delta |
|---|---|---|---|---|---|
|0|50261833995|50261849289|11245166466|11245166650|83.11957|
|1|58979382431|58979427613|11349777048|11349777590|83.361626|
|2|2591636211|2591650045|11497757814|11497757980|83.33735|
|3|15162110646|15162121922|11648603507|11648603642|83.525925|
|4|22118358237|22118370350|11732078478|11732078623|83.53793|

Intel Mac with AMD Radeon Pro 560X

| Set | cpu_timestamp0 | cpu_timestamp1 | gpu_timestamp0 | gpu_timestamp1 | cpu_delta / gpu_delta |
|---|---|---|---|---|---|
|0|10473725127789|10473725136200|3711367240|3711378120|0.77306986|
|1|10478761342978|10478761352292|8747626520|8747636600|0.92400795|
|2|10483287948145|10483287955984|13274270640|13274279920|0.8447198|
|3|10487793189051|10487793197035|17779550800|17779560080|0.8603448|
|4|10492665234848|10492665242760|22651638880|22651648800|0.79758066|

### CODE REVIEW

File: wgpu-hal/src/metal/mod.rs [hunk 1]
```diff
@@ -388,8 +390,22 @@ impl crate::Queue<Api> for Queue {
     }
 
     unsafe fn get_timestamp_period(&self) -> f32 {
-        // TODO: This is hard, see https://github.com/gpuweb/gpuweb/issues/1325
-        1.0
+        if self.shared.private_caps.supports_timestamp_period {
```
Comment (cwfitzgerald) [inline wgpu-hal/src/metal/mod.rs:393]: We should flip this to be early out

File: wgpu-hal/src/metal/mod.rs [hunk 2]
```diff
@@ -388,8 +390,22 @@ impl crate::Queue<Api> for Queue {
...
+            use objc::{msg_send, runtime::Object, sel, sel_impl};
+            let timestamps = [std::ptr::null_mut::<Object>(); 4];
+            let device = self.shared.device.lock().as_ptr();
+            let () = msg_send![device, sampleTimestamps:&timestamps[0] gpuTimestamp:&timestamps[1]];
```
Comment (cwfitzgerald) [inline wgpu-hal/src/metal/mod.rs:397]: Maybe what we could do is grab the first sample as soon as we get the device, store it, and then take a new sample in this function so that we get the longest possible sample period to get rid of the amount of noise in the reading.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]

File: wgpu-hal/src/metal/mod.rs (hunk 2 above)
Comment (jinleili) [reply wgpu-hal/src/metal/mod.rs:397]: I'm not sure if I understand it correctly, do you mean like this?
```rust
if !self.shared.private_caps.supports_timestamp_period {
   1.0
 } else {
	use objc::{msg_send, runtime::Object, sel, sel_impl};
    let timestamps0 = [std::ptr::null_mut::<Object>(); 2];
    let device = self.shared.device.lock().as_ptr();
    let () = msg_send![device, sampleTimestamps:&timestamps0[0] gpuTimestamp:&timestamps0[1]];
    if timestamps0[0] as u64 == 0 || timestamps0[1] as u64 == 0 {
        1.0
    } else {
        let timestamps1 = [std::ptr::null_mut::<Object>(); 2];
        let () = msg_send![device, sampleTimestamps:&timestamps1[0] gpuTimestamp:&timestamps1[1]];
        let cpu_delta = timestamps1[0] as u64 - timestamps0[0] as u64;
        let gpu_delta = timestamps1[1] as u64 - timestamps0[1] as u64;
        cpu_delta as f32 / gpu_delta as f32
    }         
}
```

Comment (cwfitzgerald) [reply wgpu-hal/src/metal/mod.rs:397]: (I think you replied to the wrong message 😄)

I meant something like

```rust
use objc::{msg_send, runtime::Object, sel, sel_impl};
if !self.shared.private_caps.supports_timestamp_period {
   return 1.0;
 } 

let timestamps0 = [std::ptr::null_mut::<Object>(); 2];
let device = self.shared.device.lock().as_ptr();
let () = msg_send![device, sampleTimestamps:&timestamps0[0] gpuTimestamp:&timestamps0[1]];
if timestamps0[0] as u64 == 0 || timestamps0[1] as u64 == 0 {
    return 1.0;
} 

let timestamps1 = [std::ptr::null_mut::<Object>(); 2];
let () = msg_send![device, sampleTimestamps:&timestamps1[0] gpuTimestamp:&timestamps1[1]];
let cpu_delta = timestamps1[0] as u64 - timestamps0[0] as u64;
let gpu_delta = timestamps1[1] as u64 - timestamps0[1] as u64;
cpu_delta as f32 / gpu_delta as f32
```

File: wgpu-hal/src/metal/adapter.rs [hunk 3]
```diff
@@ -202,7 +203,7 @@ impl crate::Adapter<super::Api> for super::Adapter {
                 } else if pc.msaa_desktop {
                     Tfc::SAMPLED_LINEAR
                 } else {
-                    Tfc::STORAGE
+                    Tfc::SAMPLED_LINEAR
```
Comment (kvark) [inline wgpu-hal/src/metal/adapter.rs:206]: seems unrelated to the change, could we move it to a separate PR please?

File: wgpu-hal/src/metal/mod.rs [hunk 4]
```diff
@@ -388,8 +390,23 @@ impl crate::Queue<Api> for Queue {
...
 
     unsafe fn get_timestamp_period(&self) -> f32 {
-        // TODO: This is hard, see https://github.com/gpuweb/gpuweb/issues/1325
-        1.0
+        if !self.shared.private_caps.supports_timestamp_period {
+            return 1.0;
+        }
+
+        use objc::{msg_send, runtime::Object, sel, sel_impl};
+        let timestamps0 = [std::ptr::null_mut::<Object>(); 2];
```
Comment (kvark) [inline wgpu-hal/src/metal/mod.rs:398]: should be mutable

File: wgpu-hal/src/metal/mod.rs [hunk 5]
```diff
@@ -388,8 +390,23 @@ impl crate::Queue<Api> for Queue {
...
+        let device = self.shared.device.lock().as_ptr();
+        let () = msg_send![device, sampleTimestamps:&timestamps0[0] gpuTimestamp:&timestamps0[1]];
```
Comment (kvark) [inline wgpu-hal/src/metal/mod.rs:400]: should use mutable borrows

File: wgpu-hal/src/metal/mod.rs [hunk 6]
```diff
@@ -388,8 +390,23 @@ impl crate::Queue<Api> for Queue {
...
+        if timestamps0[0] as u64 == 0 || timestamps0[1] as u64 == 0 {
+            return 1.0;
+        }
+
+        let timestamps1 = [std::ptr::null_mut::<Object>(); 2];
```
Comment (kvark) [inline wgpu-hal/src/metal/mod.rs:405]: same here

File: wgpu-hal/src/metal/mod.rs (hunk 4 above)
Comment (kvark) [inline wgpu-hal/src/metal/mod.rs:398]: why are we creating pointers to objects here? these are meant to be numbers

File: wgpu-hal/src/metal/mod.rs [hunk 7]
```diff
@@ -388,8 +390,23 @@ impl crate::Queue<Api> for Queue {
...
+        let () = msg_send![device, sampleTimestamps:&timestamps1[0] gpuTimestamp:&timestamps1[1]];
```
Comment (kvark) [inline wgpu-hal/src/metal/mod.rs:406]: shouldn't there be some kind of work between consecutive invocations of the timestamps? or do you think the resolution here is already enough?

**Review by kvark — CHANGES_REQUESTED** [review]

File: wgpu-hal/src/metal/mod.rs (hunk 4 above)
Comment (jinleili) [reply wgpu-hal/src/metal/mod.rs:398]: wgpu CI check suggested change it to this：
```sh
error: `0 as *mut _` detected
[27](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:27)
   --> wgpu-hal/src/metal/mod.rs:395:31
[28](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:28)
    |
[29](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:29)
395 |             let timestamps = [0_u64 as *mut Object; 4];
[30](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:30)
    |                               ^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null_mut::<Object>()`
[31](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:31)
    |
[32](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:32)
    = note: `-D clippy::zero-ptr` implied by `-D warnings`
[33](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:33)
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#zero_ptr
[34](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:34)

[35](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:35)
error: aborting due to previous error
[36](https://github.com/gfx-rs/wgpu/runs/5465721991?check_suite_focus=true#step:12:36)
```

File: wgpu-hal/src/metal/mod.rs (hunk 7 above)
Comment (jinleili) [reply wgpu-hal/src/metal/mod.rs:406]: Apple suggested call it at least twice, once when creating a command buffer and a second time inside the command buffer’s completion handler, but that path can't be implemented simply here, and I don't know if it makes sense to create a command buffer for that?

File: wgpu-hal/src/metal/adapter.rs (hunk 3 above)
Comment (jinleili) [reply wgpu-hal/src/metal/adapter.rs:206]: Ok :)

File: wgpu-hal/src/metal/mod.rs (hunk 5 above)
Comment (jinleili) [reply wgpu-hal/src/metal/mod.rs:400]: timestamps0 is a `[*mut Object; 2]` array, so, `&timestamps0[x]` already a mutable pointer for objc

File: wgpu-hal/src/metal/mod.rs (hunk 4 above)
Comment (jinleili) [reply wgpu-hal/src/metal/mod.rs:398]: I think I know what your mean now, will revise it

File: wgpu-hal/src/metal/mod.rs (hunk 7 above)
Comment (kvark) [reply wgpu-hal/src/metal/mod.rs]: Well, the whole `get_timestamp_period` is not in WebGPU. So we don't have to try implementing it, technically, and could come up with an entirely different API. How about no timestamp period at all? User gets some timestamps, and they can use it for comparison, but it's not really tracking absolute time.

Comment (jinleili) [reply wgpu-hal/src/metal/mod.rs]: Looks like the timestamps API is still in discussion? https://github.com/gpuweb/gpuweb/issues/2642
